Exploring V2 Features and Design (#2)
krassowski/jupyterlab-notify: README.md

JupyterLab extension to notify cell completion

![notify-extension-in-action](https://github.com/deshaw/jupyterlab-notify/blob/main/docs/notify-screenshot.png?raw=true)

> This is inspired by the notebook version [here](https://github.com/ShopRunner/jupyter-notify).

Reviewer: Let's keep it as "The initial version of this extension was inspired by ...", probably at the bottom of README.

> **Important Note**: JupyterLab Notify v2 supports `jupyter-server-nbmodel`(>= v0.1.1a2), enabling notifications to work even after the browser is closed. To enable browser-less notification support, install it with:

Reviewer: We can render it nicely with https://github.com/orgs/community/discussions/16925

```suggestion
> [!NOTE]
> JupyterLab Notify v2 supports `jupyter-server-nbmodel`(>= v0.1.1a2), enabling notifications to work even after the browser has been closed. To enable browser-less notification support, install it with:
```

```bash

> pip install jupyter-server-nbmodel

Reviewer: I think this also requires `"jupyter-docprovider>=1.0.0b1", "jupyter-server-ydoc>=1.0.0b1"` as in the `[lab]` optional group of `jupyter-server-nbmodel` ([here](https://github.com/datalayer/jupyter-server-nbmodel/blob/efc2f849e0f7dd388ac93d0c1b1fca3991885707/pyproject.toml#L32)).

Should we add it to as an optional dependency group? Then we could include the version pin for `jupyter-server-nbmodel` and the other dependencies and here just recommend:

```suggestion
pip install jupyterlab-notify[server-side-execution]
```

Author: Yes, I think this is better. Can we keep the optional dependencies as just `[server]` ? or we could name it `[server-exec]`

Author: Since the outputs aren't pushed to the UI without `jupyter-collaboration`, it would be better to keep it as an optional dependency under the `[server]` group.

Reviewer: Yes, but we do not want to depend on entire `jupyter-collaboration` but just on `jupyter-docprovider` and `jupyter-server-ydoc`.

As for naming, I think explicit is better than implicit. I would vote for `server-side-execution` but `server-exec` is fine.

```

### Configuration

> To enable Slack and email notifications, create a configuration file at `~/.jupyter/jupyterlab_notify_config.json` with the following format:

Reviewer: jupyter --
```suggestion
To enable Slack and email notifications, create a `jupyterlab_notify_config.json` configuration file under a path listed in `jupyter --paths` config section (e.g. `~/.jupyter/jupyterlab_notify_config.json`) with the following format:
```

```json
{
  "slack_token": "xoxb-your-slackbot-token",
  "slack_channel_name": "notifications",
  "email": "user@example.com"
}
```

> - `slack_token`: A valid Slack bot token with `chat:write` permissions. Refer [this article](https://help.thebotplatform.com/en/articles/7233667-how-to-create-a-slack-bot) to create your own slack bot

Reviewer: Can we point to official slack docs instead, for example this one?

```suggestion
- `slack_token`: A valid Slack bot token with `chat:write` permissions. Refer [this article](https://api.slack.com/quickstart) to create your own slack bot
```

- `slack_channel_name`: The Slack channel to post notifications to.
- `email`: The email address to receive notifications. Refer [this article](https://mailtrap.io/blog/setup-smtp-server/) to setup SMTP server.

**Note:** Ensure your JupyterLab server has SMTP access for email notifications (configured separately).

### Notification Modes

You can control when notifications are sent by setting a mode for each cell. Modes can be configured through the JupyterLab interface by clicking on the bell icon in the cell toolbar.

![image](https://github.com/user-attachments/assets/b384c0ee-88d0-47e8-9825-e42becf657a7)

**Supported modes include:**

- `always`: Sends a notification every time a cell finishes executing.
- `never`: Disables notifications for the cell.

> - `on-error`: Sends a notification only if the cell - execution fails with an error.

Reviewer: ```suggestion
- `on-error`: Sends a notification only if the cell execution fails with an error.
```

- `global-timeout`: Sends a notification if the cell execution exceeds a globally configured timeout.
- `custom-timeout`: Sends a notification if the cell execution exceeds a timeout specified for that cell.

### Global And Custom Timeout

Configure the global and custom timeout value in JupyterLab’s settings:

1. Go to Settings Editor.
2. Select notify.

> 3. Set "globalTimeout": 30 (in seconds) to apply to cells using the global-timeout mode.

Reviewer: ```suggestion
3. Set "globalTimeout": 30 (in seconds) to apply to cells using the `global-timeout` mode.
```

You can control when notifications are sent by setting a mode for each cell. Modes can be configured through the JupyterLab interface by clicking on the bell icon in the cell toolbar.

> ![image](https://github.com/darshan808/jupyterlab-notify/blob/v2-test/docs/celltoolbar-menu-screenshot.png?raw=true)

Author: I've temporarily linked to my forked branch to display the images. I'll update the links once the changes are finalized for merging.
